### DISCUSSION TITLE
machine.freq() only working after soft reboot ?

### DISCUSSION THREAD

**Discussion by GuillaumeLeclerc** [micropython/micropython#12220] [RP2040 / Pico]
Hello,

I have an audio application and I'm running a custom I2S PIO program. To be able to feed data fast enough I need to overclock the rp2040 at 240Mhz and it works fine. The problem is that it seems to only take effect until I do a soft reboot (otherwise the code is still too slow). This leads to a couple of questions:

- Is there a way to run code after a soft reboot that has been triggered programmatically:

```python
cur_freq = machine.freq()
TARGET_FREQ = 240_000_000

print("BOOTING")
if cur_freq != TARGET_FREQ:
    print("SETTING FREQ")
    machine.freq(240_000_000)
    sys.exit()
```

This code does a soft reboot however it doesn't seem to actually restart my `main.py`, which means I can't  actually have my code run at the correct frequency without calling something from the REPL.

- What exactly does sys.exit()? Maybe I could find out what is actually making my code work and do that instead of actually restarting the whole interpreter.

Thanks for the help

**Comment by GuillaumeLeclerc** [comment]
I found the issue. I was setting the frequency after initializing the PIO state machine so frequency divider was computed based off the wrong frequency.

**Comment by robert-hh** [comment]
It would be interesting what causes the effect you see. In a quick test using a MCU performance test, machine.freq() shows an immediate effect after a hard reset by a power cycle.
